roxctl/central/db/generate/generate_test.go

```diff
@@ -0,0 +1,406 @@
+				s.Assert().Contains(errOut, c.errContains)
+			} else {
+				s.Assert().Empty(errOut)
+				s.Assert().NoError(err)
+			}
+		})
+	}
+}
+
+func (s *centralDBGenerateCliTestSuite) createRootCommand(testCase testCaseType) *cobra.Command {
+	genCmd := Command(env.CLIEnvironment())
+	common.PatchPersistentPreRunHooks(genCmd)
+	if testCase.addError == downloadError {
+		genCmd.PersistentPreRunE = func(*cobra.Command, []string) error {
+			return errors.New("download error")
+		}
+	} else {
+		genCmd.PersistentPreRunE = func(*cobra.Command, []string) error {
+			cfg.SecretsByteMap = map[string][]byte{
+				"ca.pem":              []byte("ca.pem"),
+				"central-db-cert.pem": []byte("central-db-cert.pem"),
+				"central-db-key.pem":  []byte("central-db-key.pem"),
+				"central-db-password": []byte("password"),
+			}
+			return nil
+		}
+	}
+
+	s.setRunE(testCase, genCmd, "k8s", "pvc")
+	s.setRunE(testCase, genCmd, "k8s", "hostpath")
+	s.setRunE(testCase, genCmd, "k8s", "none")
+	s.setRunE(testCase, genCmd, "openshift", "pvc")
+	s.setRunE(testCase, genCmd, "openshift", "hostpath")
+	s.setRunE(testCase, genCmd, "openshift", "none")
+
+	return genCmd
+}
+
+func (s *centralDBGenerateCliTestSuite) setRunE(testCase testCaseType, root *cobra.Command, subCmds ...string) {
+	command := root
+	for _, subCmd := range subCmds {
+		command = s.lookUpCommand(command.Commands(), subCmd)
+	}
+	runE := command.RunE
+	command.RunE = func(cmd *cobra.Command, args []string) error {
+		// Run original command but tolerate the error.
+		_ = runE(cmd, args)
+		if testCase.addError == renderError {
+			return errors.New("render error")
+		}
+		s.verifyConfig(testCase, &cfg)
+		return nil
+	}
+}
+
+func (s *centralDBGenerateCliTestSuite) lookUpCommand(cmds []*cobra.Command, target string) *cobra.Command {
+	for _, cmd := range cmds {
+		if cmd.Use == target {
+			return cmd
+		}
+	}
+	s.Require().Fail("Cannot find command %s", target)
+	return nil
+}
+
+func (s *centralDBGenerateCliTestSuite) parseUsage(output string) (usage string, commands map[string][]string, flags map[string][]string) {
+	lines := strings.Split(output, "\n")
+	commands = make(map[string][]string)
+	flags = make(map[string][]string)
+	for i := 0; i < len(lines); i++ {
+		line := lines[i]
+		switch line {
+		case "Usage:":
+			i++
+			s.Assert().Less(i, len(lines))
+			usage = lines[i]
+		case "Available Commands:":
+			i++
+			s.Assert().Less(i, len(lines))
+			for ; i < len(lines) && strings.TrimSpace(lines[i]) != ""; i++ {
+				words := strings.Fields(lines[i])
+				s.Assert().Greater(len(words), 1)
+				commands[words[0]] = words[1:]
+			}
+		case "Flags:":
+			i++
+			s.Assert().Less(i, len(lines))
+			for ; i < len(lines) && strings.TrimSpace(lines[i]) != ""; i++ {
+				words := strings.Fields(lines[i])
+				s.Assert().Greater(len(words), 1)
+				flags[words[0]] = words[1:]
+			}
+		}
+	}
+	return
+}
+
+func (s *centralDBGenerateCliTestSuite) verifyConfig(c testCaseType, config *renderer.Config) {
```

Comment:
This function has a lot of logic. I think it should be clean assertion `Equal(actual, expected *renderer.Config)` and `expected` config should be part of `testCaseType`